### PR TITLE
Implement save for later

### PR DESCRIPTION
**Pull request by NathanielBennett**
Iniitial Save For Later implementation. There are a couple cfn files which can be removed post migratiion and launch

(I'll squash the commits)
@markrichardsg i thieved some of your code..

### CODE REVIEW

File: src/main/scala/com/gu/sfl/lib/Merge.scala [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import scala.reflect.internal.util.Collections.distinctBy
+
+//This is a virtual straight rob from identity
+object Merge {
+
+  implicit class RichList[A](val l: List[A]) {
+    def tailOrNil = if(l.isEmpty) Nil else l.tail
+  }
+
+  def mergeListBy[A, K](before: List[A], after: List[A])(key: A => K)(implicit ordering: Ordering[A]): List[A] = {
```
Comment (markrichardsg) [inline src/main/scala/com/gu/sfl/lib/Merge.scala]: Is this simpler?
`distinctBy(before ++ after)(key).sorted(ordering)`

File: src/main/scala/com/gu/sfl/lib/SavedArticlesMerger.scala [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    def loop(articles: SavedArticles, retries: Int): Try[Option[SyncedPrefs]] = {
+
+      if(retries == 0) {
+        logger.debug(s"Failed to merge saved articles for user: $userId")
+        Failure(SavedArticleMergeError("Conflicting version in savedArticles") )
+      } else {
+        savedArticlesPersistence.read(userId) match {
+          case Success(Some(currentArticles)) =>
+            val mergedArticles = Merge.mergeListBy(currentArticles.articles, articles.articles)(_.id)
+            if(currentArticles.version == articles.version) {
```
Comment (markrichardsg) [inline src/main/scala/com/gu/sfl/lib/SavedArticlesMerger.scala]: Is this a candidate for a conditional expression? https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Expressions.ConditionExpressions.html

File: src/main/scala/com/gu/sfl/model/model.scala [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+object SavedArticles {
+  def nextVersion : String = Instant.now().toEpochMilli.toString
+  def apply(articles: List[SavedArticle]) : SavedArticles = SavedArticles(nextVersion, articles)
+}
+
+case class SavedArticles(version: String, articles: List[SavedArticle]) extends SyncedPrefsData {
+  override def advanceVersion: SavedArticles = copy(version = nextVersion)
+  def ordered: SavedArticles = copy()
```
Comment (markrichardsg) [inline src/main/scala/com/gu/sfl/model/model.scala]: Is this supposed to be:
`def ordered: SavedArticles = copy(articles = articles.sorted)`

File: src/main/scala/com/gu/sfl/lib/Merge.scala (hunk 1 above)
Comment (NathanielBennett) [reply src/main/scala/com/gu/sfl/lib/Merge.scala]: It looks simpler, admittedly, but it breaks all the tests.

File: src/main/scala/com/gu/sfl/lib/SavedArticlesMerger.scala (hunk 2 above)
Comment (NathanielBennett) [reply src/main/scala/com/gu/sfl/lib/SavedArticlesMerger.scala]: Almost, but not really. We're dealing with a scenario where a device is saving to a record which has been updated by another device ( ie, the version in the current request is different from the currently persisted version ) where the version is the same we know this isn't case. Where it differs, we take the currently saved record and attemp to merge the article list on the request with the list currently persisted. We then try to to save the merged list with the currently (as in when we started merging ) persisted version value. If we do a conditional update, we can't really do this

File: src/test/scala/com/gu/sfl/lib/MergeTest.scala [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    "perform a trivial merge" in {
+       mergeList(List(1,2,3), List(1,2,3)).toSet shouldEqual(Set(1,2,3))
+    }
+
+    "merge from an empty list to a new target" in {
+      mergeList(Nil, List(1,2,3)).toSet shouldEqual(Set(1,2,3))
+    }
+
+    "merge from a full delete" in {
+      mergeList(List(1,2,3), Nil) shouldEqual(Nil)
```
Comment (markrichardsg) [inline src/test/scala/com/gu/sfl/lib/MergeTest.scala]: Is this a merge?

File: src/main/scala/com/gu/sfl/lib/Merge.scala (hunk 1 above)
Comment (markrichardsg) [reply src/main/scala/com/gu/sfl/lib/Merge.scala]: See review commended added to tests.
Perhaps merge isn't the intended behaviour and the tests need revising or this isn't a "merge" - could the function be renamed if this isn't a straight forward merge?

**Pull request merged**